// @unibeautify/cli/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const unibeautify_1 = require("unibeautify");
const requireg = require("requireg");
const gSearch = require("g-search");
/**
Find all globally installed beautifiers.
*/
function findInstalledBeautifiers() {
    return new Promise((resolve, reject) => {
        gSearch()
            .then((globalPackages) => {
            const packageNames = globalPackages.map(pkg => pkg.name);
            const beautifierNames = packageNames.filter(isBeautifierPackageName);
            return resolve(beautifierNames);
        })
            .catch((error) => {
            return reject(error);
        });
    });
}
exports.findInstalledBeautifiers = findInstalledBeautifiers;
function isBeautifierPackageName(packageName) {
    return /beautifier-.*/.test(packageName);
}
exports.isBeautifierPackageName = isBeautifierPackageName;
function loadBeautifiers(beautifierNames) {
    const beautifiers = beautifierNames
        .map(beautifierName => {
        try {
            return requireg(beautifierName).default;
        }
        catch (error) {
            console.error(error);
            return null;
        }
    })
        .filter(Boolean);
    return unibeautify_1.default.loadBeautifiers(beautifiers);
}
exports.loadBeautifiers = loadBeautifiers;
function loadInstalledBeautifiers() {
    return findInstalledBeautifiers().then((beautifierNames) => {
        return loadBeautifiers(beautifierNames);
    });
}
exports.loadInstalledBeautifiers = loadInstalledBeautifiers;
function setupUnibeautify() {
    return loadInstalledBeautifiers();
}
exports.setupUnibeautify = setupUnibeautify;
function getSupportedLanguages() {
    return unibeautify_1.default.supportedLanguages.map(language => {
        return language.name;
    });
}
exports.getSupportedLanguages = getSupportedLanguages;
function getAllLanguages() {
    return unibeautify_1.default.getLoadedLanguages().map(language => {
        return language.name;
    });
}
exports.getAllLanguages = getAllLanguages;
